pow/src/main/java/tech/pegasys/artemis/pow/DepositProcessingController.java

```diff
@@ -0,0 +1,210 @@
+    final BigInteger toBlock;
+    final BigInteger fromBlock;
+
+    if (active || latestCanonicalBlockNumber.equals(latestSuccessfullyQueriedBlock)) {
+      return;
+    }
+    active = true;
+
+    fromBlock = latestSuccessfullyQueriedBlock.add(BigInteger.ONE);
+    toBlock = latestCanonicalBlockNumber;
+
+    depositFetcher
+        .fetchDepositsInRange(fromBlock, toBlock)
+        .finish(
+            __ -> onSubscriptionDepositRequestSuccessful(toBlock),
+            (err) -> onSubscriptionDepositRequestFailed(err, fromBlock, toBlock));
+  }
+
+  private synchronized void fetchLatestDepositsOneBlockAtATime() {
+    final BigInteger nextBlockNumber;
+
+    synchronized (DepositProcessingController.this) {
+      if (active || latestCanonicalBlockNumber.equals(latestSuccessfullyQueriedBlock)) {
+        return;
+      }
+      active = true;
+
+      nextBlockNumber = latestSuccessfullyQueriedBlock.add(BigInteger.ONE);
+    }
+
+    depositFetcher
+        .fetchDepositsInRange(nextBlockNumber, nextBlockNumber)
+        .thenCompose(
+            __ ->
+                eth1Provider.getGuaranteedEth1BlockFuture(
+                    UnsignedLong.valueOf(nextBlockNumber), asyncRunner))
+        .thenAccept(
+            block -> {
+              final BigInteger blockNumber = block.getNumber();
+              LOG.trace("Successfully fetched block {} for min genesis checking", blockNumber);
+              LOG.trace(
+                  "Seconds until min genesis block {}",
+                  Constants.MIN_GENESIS_TIME.minus(
+                      calculateCandidateGenesisTimestamp(block.getTimestamp())));
+              if (MinimumGenesisTimeBlockFinder.compareBlockTimestampToMinGenesisTime(block) >= 0) {
+                notifyMinGenesisTimeBlockReached(eth1EventsChannel, block);
+                isBlockByBlockModeOn = false;
+                LOG.debug("Switching back to fetching deposits by range");
+              }
+            })
+        .finish(
+            __ -> onSubscriptionDepositRequestSuccessful(nextBlockNumber),
+            (err) -> onSubscriptionDepositRequestFailed(err, nextBlockNumber));
+  }
+
+  private synchronized void onSubscriptionDepositRequestSuccessful(BigInteger requestToBlock) {
+    active = false;
+    latestSuccessfullyQueriedBlock = requestToBlock;
+
+    if (latestCanonicalBlockNumber.compareTo(latestSuccessfullyQueriedBlock) > 0) {
+      fetchLatestSubscriptionDeposits();
+    }
+  }
+
+  private synchronized void onSubscriptionDepositRequestFailed(
+      Throwable err, BigInteger fromBlock) {
+    onSubscriptionDepositRequestFailed(err, fromBlock, fromBlock);
+  }
+
+  private synchronized void onSubscriptionDepositRequestFailed(
+      Throwable err, BigInteger fromBlock, BigInteger toBlock) {
+    active = false;
+
+    if (!fromBlock.equals(toBlock)) {
+      LOG.warn(
+          "Failed to fetch deposit events for block numbers in the range ({}, {})",
```

Comment:
Add "Retrying" to the end so the user knows it's being reattempted.

Reply:
Done.